zip-cli binary crate

cosmicexplorer opened the pull request:
# Problem
We often want to generate zip files for use in testing and benchmarking, as in #208 and #233. In #233 I added a python script in `tests/data/` which uses the built-in `zipfile` stdlib module to create zip files. While this works, it also seems a little silly when we've done most of the work already in this crate to create a highly flexible interface to zip files. Separately, it seems like having a CLI to work on would be a good way to learn more about ways we can make this crate work better with filesystems (e.g. in a TODO I note that we should probably modify `.add_symlink_from_path()` to accept an `OsStr` target).

## Prior Investigations
I previously created https://github.com/cosmicexplorer/medusa-zip as a sandbox to mess around with high-performance zip merging and splitting. That was partially because the `zip` crate was unmaintained at the time, and partially because I was under the impression that some of the techniques would be difficult to retrofit into a standard zip library. It was also an excuse to play around with pyo3 and learn how to build for many python platforms with cibuildwheels.

Since then, I've found that:
(1) The use of `async` as in `medusa-zip` tends to heavily slow down programs during purely filesystem-based I/O, as they tend not to wait for very long (this is why linux does not offer very many nonblocking file APIs). `async` tends to be harmful to latency and more useful for very high *throughput*, especially when waiting on network latencies.
(2) High levels of parallelism do not require egregious usage of temporarily output files as in `medusa-zip`, but can be implemented with `pread()` and pipes on `#[cfg(unix)]` targets as in #208.

# Solution
Create a `zip-cli` crate in the `cli/` subdir. This isn't published anywhere yet, as the main purpose right now is to make it easier to generate test data for benchmarks. This is planned to provide three subcommands:
- `compress`: accept a specification of entries and options and generate a zip file output, like `zip`.
- `info`: write out info about a zip file, like `zipinfo`.
- `extract`: extract individual entries or an entire archive to stdout or the filesystem, like `unzip`.

# Result
Currently, only `compress` is implemented, but it seems to work great:
```shell
> cargo run -p zip-cli -- compress -h
   Compiling zip-cli v0.0.0 (/home/cosmicexplorer/tools/zip/cli)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.32s
     Running `target/debug/zip-cli compress -h`
do a compress

Usage: zip-cli compress [OPTIONS] [positional-path]...

Options:
  -o, --output-file <output-file>  Output zip file path to write.
      --stdout                     Allow writing output to stdout even if stdout is a tty.
  -h, --help                       Print help (see more with '--help')

ATTRIBUTES:
  -c, --compression-method <compression-method>
          Which compression technique to use.
          Defaults to deflate if not specified. [possible values: stored, deflate, deflate64, bzip2, zstd]
  -l, --compression-level <compression-level>
          How much compression to perform, from 0..=24.
  -m, --mode <mode>
          Unix permissions to apply to the file, in octal (like chmod).
      --large-file <large-file>
          Whether to enable large file support. [possible values: true, false]

ENTRIES:
  -n, --name <name>
          The name to apply to the entry.
  -d, --dir
          Create a directory entry.
  -s, --symlink
          Make the next entry into a symlink entry.
  -i, --imm <immediate>
          Write an entry containing this data.
  -f, --file <file>
          Write an entry with the contents of this file path.
  -r, --recursive-dir <recursive-directory>
          Write all the recursive contents of this directory path.
  [positional-path]...
          Write the file or recursive directory contents, relativizing the path.
> cargo run -p zip-cli -- -v compress -o out.zip -n a -d -n b -s -i c -m755 -f tmp/file.txt 
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.04s
     Running `target/debug/zip-cli -v compress -o out.zip -n a -d -n b -s -i c -m755 -f tmp/file.txt`
writing compressed zip to output file path "out.zip"
default zip entry options: FileOptions { compression_method: Deflated, compression_level: None, last_modified_time: DateTime::from_date_and_time(2024, 8, 12, 16, 42, 29)?, permissions: None, large_file: false, encrypt_with: None, extended_options: (), alignment: 1, zopfli_buffer_size: Some(32768) }
setting name of next entry to "a"
writing dir entry
setting name of next entry to "b"
setting symlink flag for next entry
writing immediate symlink entry with name "b" and target "c"
setting file mode 0o755
writing file entry from path "tmp/file.txt" with name "tmp/file.txt"
> unzip -l out.zip
Archive:  out.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
        0  2024-08-12 16:42   a/
        1  2024-08-12 16:42   b
        4  2024-08-12 16:42   tmp/file.txt
---------                     -------
        5                     3 files
```


# TODO
`medusa-zip` accepts a JSON interface for specifying inputs to compress, which is less useful for the shell but much more useful for execution e.g. from build tools. Having all the inputs planned out in advance also makes it possible to perform extremely simple and powerful optimizations by creating "sub-zips" in parallel and merging them without decompressing. This is something I plan to propose after I spend more time on #208.

Pr0methean commented on Cargo.toml:
```diff
@@ -22,8 +23,19 @@ build = "src/build.rs"
+[workspace]
+members                 = [
+    ".",
+    "cli",
```
Should `fuzz` also be a member?

cosmicexplorer replied:
This is a great idea! Unfortunately, I get this incredibly strange error during linking:
```shell
error: linking with `cc` failed: exit status: 1
  |
  = note: LC_ALL="C" PATH="/home/cosmicexplorer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin:/home/cosmicexplorer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/self-contained:/home/cosmicexplorer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin:/home/cosmicexplorer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/self-contained:/home/cosmicexplorer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin:/home/cosmicexplorer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/self-contained:/home/cosmicexplorer/bazel-bullshit/:/usr/lib/jvm/default-runtime/bin:/home/cosmicexplorer/.pyenv/shims:/home/cosmicexplorer/tools/emacs/rex-install/bin:/home/cosmicexplorer/.local/bin:/home/cosmicexplorer/go/bin:/home/cosmicexplorer/.local/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/usr/lib/rustup/bin:/usr/bin/core_perl:/usr/bin/vendor_perl:/home/cosmicexplorer/.zsh/snippets/bash:/home/cosmicexplorer/go/bin:/home/cosmicexplorer/.cargo/bin" VSLANG="1033" "cc" "-m64" "/tmp/rustcgcg83u/symbols.o" "-Wl,-Bstatic" "-Wl,--whole-archive" "-Wl,/home/cosmicexplorer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc-nightly_rt.asan.a" "-Wl,--no-whole-archive" "/home/cosmicexplorer/tools/zip/target/x86_64-unknown-linux-gnu/release/deps/fuzz_read-73bbb0616ed32380.fuzz_read.775828843f50ae8f-cgu.0.rcgu.o" "-Wl,--as-needed" "-L" "/home/cosmicexplorer/tools/zip/target/x86_64-unknown-linux-gnu/release/deps" "-L" "/home/cosmicexplorer/tools/zip/target/release/deps" "-L" "/home/cosmicexplorer/tools/zip/target/x86_64-unknown-linux-gnu/release/build/libfuzzer-sys-394e55403fa44217/out" "-L" "/home/cosmicexplorer/tools/zip/target/x86_64-unknown-linux-gnu/release/build/tikv-jemalloc-sys-9cd5b3f80b1e0b3c/out/build/lib" "-L" "/usr/lib" "-L" "/home/cosmicexplorer/tools/zip/target/x86_64-unknown-linux-gnu/release/build/zstd-sys-28dcdb0894e5b147/out" "-L" "/home/cosmicexplorer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/tmp/rustcgcg83u/libzstd_sys-f8986c36e00f9a6f.rlib" "/tmp/rustcgcg83u/libtikv_jemalloc_sys-b8ade72b5ff520e1.rlib" "/tmp/rustcgcg83u/liblibfuzzer_sys-64607a53ace2a9b6.rlib" "/home/cosmicexplorer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-4d7d16bbf0636a40.rlib" "-Wl,-Bdynamic" "-lbz2" "-lstdc++" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-B/home/cosmicexplorer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/gcc-ld" "-B/home/cosmicexplorer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/gcc-ld" "-B/home/cosmicexplorer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/gcc-ld" "-fuse-ld=lld" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/home/cosmicexplorer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-L" "/home/cosmicexplorer/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/self-contained" "-o" "/home/cosmicexplorer/tools/zip/target/x86_64-unknown-linux-gnu/release/deps/fuzz_read-73bbb0616ed32380" "-pie" "-Wl,-z,relro,-z,now" "-Wl,--strip-all" "-nodefaultlibs"
# ... (snip)
          rust-lld: error: undefined symbol: __sancov_gen_.51
          >>> referenced by fuzz_read.775828843f50ae8f-cgu.0
          >>>               /home/cosmicexplorer/tools/zip/target/x86_64-unknown-linux-gnu/release/deps/fuzz_read-73bbb0616ed32380.fuzz_read.775828843f50ae8f-cgu.0.rcgu.o:(asan.module_dtor.10819)
          
          rust-lld: error: undefined symbol: __sancov_gen_.1086
          >>> referenced by fuzz_read.775828843f50ae8f-cgu.0
          >>>               /home/cosmicexplorer/tools/zip/target/x86_64-unknown-linux-gnu/release/deps/fuzz_read-73bbb0616ed32380.fuzz_read.775828843f50ae8f-cgu.0.rcgu.o:(asan.module_dtor.10892)
          
          rust-lld: error: too many errors emitted, stopping now (use --error-limit=0 to see all errors)
          collect2: error: ld returned 1 exit status
          
error: could not compile `zip-fuzz` (bin "fuzz_read") due to 1 previous error
Error: failed to build fuzz script: ASAN_OPTIONS="detect_odr_violation=0" RUSTFLAGS="-Cpasses=sancov-module -Cllvm-args=-sanitizer-coverage-level=4 -Cllvm-args=-sanitizer-coverage-inline-8bit-counters -Cllvm-args=-sanitizer-coverage-pc-table -Cllvm-args=-sanitizer-coverage-trace-compares --cfg fuzzing -Clink-dead-code -Zsanitizer=address -Cllvm-args=-sanitizer-coverage-stack-depth -Cdebug-assertions -C codegen-units=1" "cargo" "build" "--manifest-path" "/home/cosmicexplorer/tools/zip/fuzz/Cargo.toml" "--target" "x86_64-unknown-linux-gnu" "--release" "--config" "profile.release.debug=true" "--all-features" "--bin" "fuzz_read"
```

This is from:
1. putting `"fuzz"` in `Cargo.toml`'s `workspace.members` array.
2. commenting out the `workspace.members` array in `fuzz/Cargo.toml`.

This seems to happen pretty reliably, and I honestly have no clue how to debug it. It would very much be nice to be able to reuse the `arbitrary` dependency declaration. If I dug more into how `cargo fuzz` is executed and configured I could probably figure this out.

cosmicexplorer replied:
Ahhhh, this is because of our `strip` and other stuff in `profile.release`! I think I can turn that off for fuzz targets.

cosmicexplorer replied:
Ugh, ok. It looks like `cargo fuzz` is *strictly* only able to use `release` (by default) or `dev` (with a command-line flag). [Custom Profiles](https://doc.rust-lang.org/cargo/reference/profiles.html#custom-profiles) might be appropriate. I don't really know how people distribute rust binaries, though--do they just run `cargo install zip-cli` and it builds from source? What profile would that use? I'll figure that out.

cosmicexplorer replied:
So it looks like we *could* e.g. tell users "run `cargo install --profile release-lto zip-cli`" for a smaller binary, maybe. I suspect there are best practices already established for this sort of thing that we should look to, since we're surely not the only person who has wanted to put a fuzz crate and a binary crate in the same workspace. I will now take a look at your other comment regarding splitting up the binary.

cosmicexplorer replied:
Sorry, to clarify:
1. `cargo fuzz` is only able to use the `dev` or `release` profiles (`release` by default)
    - I assume we want to use `release` because fuzzing is slow, so we don't really have an option to change anything there -- we're just stuck with fuzzing using `profile.release`.
2. turning on `lto = false` in `profile.release` breaks `cargo fuzz` (some linker error about ASan, linked above)
   - `lto = true` converts our current 1.6MB binary to 1.3MB, which is good but not as good as `strip`, which *does* work just fine.
3. [Custom profiles](https://doc.rust-lang.org/cargo/reference/profiles.html#custom-profiles) can inherit from `profile.release` and additionally set e.g. `lto = true`, without interfering with other profiles.
    - For local development, we can run `cargo build --profile release-lto`. 
4. `cargo install` *also* uses `profile.release` by default, and users generally expect to be able to just run `cargo install zip-cli` without additional flags.
    - However, `cargo install` at least *does* support custom profiles. So we *could* tell users "please run `cargo install --profile release-lto` for a smaller binary."

Pr0methean replied:
Actually, I suspect that disabling `aes-crypto` and compression algos other than `deflate` will make a bigger difference to the binary's size and memory footprint than any combination of `lto` and `strip`, and still cover the most common use cases.

Pr0methean commented on cli/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+[package]
+name = "zip-cli"
+version = "0.0.0"
```
Change this to `0.0.1` for alpha or `0.1.0` for beta.

cosmicexplorer replied:
Done!

Pr0methean commented on cli/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+repository = "https://github.com/zip-rs/zip2.git"
+keywords = ["zip", "archive", "compression", "cli"]
+categories = ["command-line-utilities", "compression", "filesystem", "development-tools::build-utils"]
+rust-version = "1.73.0"
```
Increase this in order to update `clap`.

cosmicexplorer replied:
Done!

Pr0methean commented on cli/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+
+[[bin]]
+name = "zip-cli"
+
```
Can we split this into 2 `bin` targets, a "lite" one (no features except `deflate-flate2` and one of the `flate2` backends, opt-level = "z") and a "full" one (all compression algorithms, plus `aes-crypto` and either `chrono` or `time`, with the usual optimization for speed given abundant memory)?

Pr0methean commented on cli/src/args.rs:
```diff
@@ -0,0 +1,475 @@
+    /// with deflate (default)
+    Deflate,
+    /// with deflate64
+    Deflate64,
```
Gate `Deflate64`, `Bzip2` and `Zstd` on their respective features.

cosmicexplorer replied:
Done!

Pr0methean reviewed (changes requested):
This is just a partial review; I'll review most of the code changes tomorrow or Wednesday.

cosmicexplorer commented:
Ok, after all those very long replies, the solution is actually quite simple:
- make `zip-cli` expose a `[lib]` target as well as a `[[bin]]` target
  - forward all features from `zip` in its own `[features]` table
- make another crate `zip-clite` (in its own workspace) that depends on `zip-cli` as a library (to perform clap argument parsing and implement each subcommand)

Result:
```shell
> cd cli
> cargo build --release
> ls target/release/zip-cli
-rwxr-xr-x 2 cosmicexplorer wheel 1.8M Aug 13 09:34 target/release/zip-cli*
> cd clite
> cargo build --release
> ls target/release/zip-clite
-rwxr-xr-x 2 cosmicexplorer wheel 825k Aug 13 09:35 target/release/zip-clite*
```
So we end up having to publish two separate crates, but (1) this allows us to achieve the precise separation you described in https://github.com/zip-rs/zip2/pull/235#discussion_r1714535133, and (2) this also means users can actually just run `cargo install zip-cli` and get the thing they want, or `cargo install zip-clite` if they want the smaller version.

Currently `zip-clite` does not expose any features of its own (it hardcodes `deflate-flate2` and `deflate-zlib`), but `zip-cli` does, and it can be narrowed down to a smaller binary as desired:
```shell
> cd cli
> cargo build --release --no-default-features --features deflate-zlib,bzip2,lzma,xz
> ls target/release/zip-cli                                                          
-rwxr-xr-x 2 cosmicexplorer wheel 1.2M Aug 13 09:40 target/release/zip-cli*
```
Note that this 1.2M is smaller than the previous 1.8M with default features enabled.

I like this solution a lot and I'm really glad we didn't have to delve into anything more complex!

cosmicexplorer commented:
If we really wanted to care about artifact size for `clite` (do we?), we could consider removing `clap`, especially since the `compress` subcommand parses its order-dependent arguments in a relatively idiosyncratic way--we essentially have to work around clap a lot (see this incredibly lengthy bit of boilerplate we have to do to express our order-dependent args in clap: https://github.com/zip-rs/zip2/blob/78c659bf0c46507f075565cf9f20c16c11755812/cli/src/args.rs#L83-L317; that is a 200-line function).

Especially if our goal here is more about providing an interface to the `zip` library rather than maintaining a carefully-crafted CLI that users employ constantly in day-to-day work, I think we may be within our rights to consider parsing args outside of clap. I may try to see how easy it is to drop clap from this and whether it feels like the right idea.

cosmicexplorer commented:
Ugh, I actually really like the help text generation mechanism for clap and I don't think binary size is that much of a concern for us at all. I think we definitely have an excuse to go outside of clap if we feel like it but I would prefer to work on other things right now.

Pr0methean commented:
Have you looked at the other parsers listed here? https://rust-cli-recommendations.sunshowers.io/cli-parser.html#alternatives-to-clap

Pr0methean commented:
It also occurs to me that it might be helpful to feature-gate the dependency on `clap`; for example, if `clite` was being deployed for shell-script use to a server where you couldn't SSH in without beeping someone's pager and getting the 33rd degree -- as is true for most prod servers at AWS -- then the help text wouldn't matter.

cosmicexplorer commented:
Thanks so much for that discussion of use cases for this binary!! It is a new day and I think it is actually not a bad idea to try parsing args by hand! Will report back ^_^!

cosmicexplorer commented:
Seems pretty clear that clap is unnecessary, and I've removed it (the parsing is also much easier to follow, since clap requires immense circumlocutions to support...an ordered sequence of args).

I'd also like to note: I'm currently depending on [`color-eyre`](https://docs.rs/color-eyre/latest/color_eyre/) in `zip-cli` and just plain [`eyre`](https://docs.rs/eyre/latest/eyre/) in `zip-clite`. Right now, the sizes are 1.4M and 477k in `--release` mode:
```shell
> cd cli
> ls target/release/zip-cli
-rwxr-xr-x 2 cosmicexplorer wheel 1.4M Aug 14 14:07 target/release/zip-cli*
> cd clite
> ls target/release/zip-clite 
-rwxr-xr-x 2 cosmicexplorer wheel 477k Aug 14 14:07 target/release/zip-clite*
```

For iterating on this, `color-eyre` is vaguely useful, but I don't think we'll actually need a backtrace here, since the error cases are either going to be argument parsing or i/o errors, where we should be providing a formatted error message and exiting with a specific error code instead of printing a backtrace. As with the decision to drop `clap`, we're not making a big hulking command-line application with so many flags and failure modes we can't remember them all. I *think* that for every usage of `?` in the app right now, we should be able to provide a specific and useful error message. For i/o errors especially, we don't actually get any benefit from `eyre`/`anyhow`, since they don't capture the path of the file that produced the error for us. So we need to manually handle those errors anyway to produce a good error message. I don't think this will be incredibly difficult or time-consuming.

cosmicexplorer commented:
Hm, I tried removing `eyre` and `color-eyre` respectively, and actually only got 1.4M and 465k respectively for `cargo build --release` (so `zip-cli` stayed the same size, and `zip-clite` only became 12kb smaller). So what I might do then is to keep them in, and make use of [`WrapErr`](https://docs.rs/eyre/latest/eyre/trait.WrapErr.html) to provide context along with the error messages, so we can still get that context we need to make good error messages--but no reason to remove the nice error handling/propagation crates if it doesn't help our binary size (and better error reporting upon failure seems useful enough to keep, and `eyre` seems safe and widely used).

cosmicexplorer commented:
Ok, after doing a lot of `.wrap_err_with()` calls, we're now up to 1.7M for `zip-cli` and 780k for `zip-clite`. Error messages look like this:
```shell
> cargo run -- -v compress -o out.zip -n a -d -n b -s -i c -m 755 -f tmp/file2.txt
writing compressed zip to output file path "out.zip"
default zip entry options: FileOptions { compression_method: Deflated, compression_level: None, last_modified_time: DateTime::default(), permissions: None, large_file: false, encrypt_with: None, extended_options: (), alignment: 1, zopfli_buffer_size: Some(32768) }
setting name of next entry to "a"
writing dir entry
setting name of next entry to "b"
setting symlink flag for next entry
writing immediate symlink entry with name "b" and target "c"
setting file mode 0o755
writing file entry from path "tmp/file2.txt" with name "tmp/file2.txt"
Error: 
   0: tmp/file2.txt
   1: No such file or directory (os error 2)

Location:
   /home/cosmicexplorer/tools/zip/cli/src/compress.rs:286

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
```

I was able to mimic this somewhat without the `eyre` crate, and it did work (although it required a lot more boilerplate, which is somewhat of a risk in itself due to poor readability), so let me know how you feel about error crates like eyre.

Note that invalid cli args are processed without any error crate, and we explicitly call `process::exit()` ourselves:
```shell
> cargo run -- -v compress -o out.zip -n a -d -n b -s -s -i c -m 755 -f tmp/file2.txt
writing compressed zip to output file path "out.zip"
default zip entry options: FileOptions { compression_method: Deflated, compression_level: None, last_modified_time: DateTime::from_date_and_time(2024, 8, 14, 20, 2, 54)?, permissions: None, large_file: false, encrypt_with: None, extended_options: (), alignment: 1, zopfli_buffer_size: Some(32768) }
setting name of next entry to "a"
writing dir entry
setting name of next entry to "b"
setting symlink flag for next entry
setting symlink flag for next entry
error: symlink flag provided twice before entry

Usage: target/debug/zip-cli compress [-h|--help] [OUTPUT-FLAG] [ENTRIES]... [--] [PATH]...
```

This is what the `compress` command help output looks like. Note that the possible values for `-c/--compression-method` are gated based on the selected features:
```shell
> cargo run -- compress --help
do a compress

Usage: target/debug/zip-cli compress [-h|--help] [OUTPUT-FLAG] [ENTRIES]... [--] [PATH]...

  -h, --help    Print help

Output flags:
Where and how to write the generated zip archive.
  -o, --output-file <file>
          Output zip file path to write.
          The output file is currently always truncated if it already exists.
          If not provided, output is written to stdout.

      --stdout
          Allow writing output to stdout even if stdout is a tty.

ENTRIES:
After at most one output flag is provided, the rest of the command line is attributes and
entry data. Attributes modify later entries.

Sticky attributes:
These flags apply to everything that comes after them until reset by another instance of the
same attribute. Sticky attributes continue to apply to positional arguments received after
processing all flags.

  -c, --compression-method <method-name>
          Which compression technique to use.
          Defaults to deflate if not specified.

          Possible values:
          - stored:    uncompressed
          - deflate:   with deflate (default)
          - deflate64: with deflate64
          - bzip2:     with bzip2
          - zstd:      with zstd


  -l, --compression-level <level>
          How much compression to perform, from 0..=24.
          The accepted range of values differs for each technique.

  -m, --mode <mode>
          Unix permissions to apply to the file, in octal (like chmod).

      --large-file [true|false]
          Whether to enable large file support.
          This may take up more space for records, but allows files over 32 bits in length to be
          written, up to 64 bit sizes.

Non-sticky attributes:
These flags only apply to the next entry after them, and may not be repeated.

  -n, --name <name>
          The name to apply to the entry.

  -s, --symlink
          Make the next entry into a symlink entry.
          A symlink entry may be immediate with -i, or it may read the symlink value from the
          filesystem with -f.

Entry data:
Each of these flags creates an entry in the output zip archive.

  -d, --dir
          Create a directory entry.
          A name must be provided beforehand with -n.

  -i, --imm <immediate>
          Write an entry containing this data.
          A name must be provided beforehand with -n.

  -f, --file <path>
          Write an entry with the contents of this file path.
          A name may be provided beforehand with -n, otherwise the name will be inferred from
          relativizing the given path to the working directory.

  -r, --recursive-dir <dir>
          Write all the recursive contents of this directory path.
          A name may be provided beforehand with -n, which will be used as the prefix for all
          recursive contents of this directory. Otherwise, the name will be inferred from
          relativizing the given path to the working directory.

Positional entries:
  [PATH]...
          Write the file or recursive directory contents, relativizing the path.
          If the given path points to a file, then a single file entry will be written.
          If the given path is a symlink, then a single symlink entry will be written.
          If the given path refers to a directory, then the recursive contents will be written.
```

cosmicexplorer commented:
> Cool! Can `--large-file` possibly be set automatically when the input isn't a pipe or socket or open for writing?

Yes!!! (In fact it currently doesn't support anything except regular files, directories, and symlinks, although it definitely could). We can keep the existing "sticky" behavior for explicit `--large-file [true|false]` flags.

cosmicexplorer commented:
Decided that `info` is probably simple enough to implement now, but I want to get #208 in first before doing `extract` (I had been letting it sit but I can see you gave me some great feedback), and think it would be useful for benchmark creation to have this on master instead of waiting around for python zipfile. I just made the description for `info` and `extract` say TODO for now (and fleshed out the `compress` description); let me know if there's any more functionality it would be useful to have at this point.

I almost want to remove the `--large-file` option after implementing your suggestion to just infer it for file inputs, but I think it would be reasonable to support pipes and sockets in the future and it seems like a good idea to require the user to explicitly specify that instead of accidentally trying to compress a massive pipe.

Pr0methean commented:
> Decided that info is probably simple enough to implement now, but I want to get https://github.com/zip-rs/zip2/pull/208 in first before doing extract

Then I guess I'll wait for that PR to be merged before I spend much more energy on this one.

Pr0methean reviewed (changes requested):
Another partial review.

cosmicexplorer commented on cli/clite/Cargo.toml:
```diff
@@ -0,0 +1,36 @@
+[profile.release]
+strip = true
+lto = true
+opt-level = "s"
```
It's actually larger!

```shell
# with opt-level = "s"
> ls target/release/zip-clite      
-rwxr-xr-x 2 cosmicexplorer wheel 780k Aug 14 15:58 target/release/zip-clite*
# with opt-level = "z"
> ls target/release/zip-clite
-rwxr-xr-x 2 cosmicexplorer wheel 801k Aug 17 13:47 target/release/zip-clite*
```

I also think there is a really good argument for *not* depending on `color-eyre` especially for the `clite` binary, since adding usages of `.wrap_err_with(...)` ended up adding a few hundred kb to the binary. I'm going to see if:
1. I can keep the same code but use the default `eyre` crate and get an improvement,
2. I can replace it with `anyhow` (and give up nice colored errors) and get an improvement,
3. I can do like I did with `clap` and produce my own tiny `WrapErr` trait and associated `Error` object that avoids calling into other libraries.

I think as with `clap`, since we're not making a really hulking server application with tons of error cases, (3) is not actually a bad idea. When I tried something like (3) previously, I didn't try making something like a `WrapErr` trait, but methods I had to call manually which was rather annoying and difficult to extend. I think doing it that way is probably a good idea, honestly.

cosmicexplorer commented on cli/clite/src/main.rs:
```diff
@@ -0,0 +1,26 @@
+    };
+
+    match command {
+        ZipCommand::Info => eyre::bail!("info command not implemented"),
```
That sounds great!!

cosmicexplorer commented:
> > Decided that info is probably simple enough to implement now, but I want to get #208 in first before doing extract
> 
> Then I guess I'll wait for that PR to be merged before I spend much more energy on this one.

Thanks so much for these thoughtful comments, I really appreciate it! Really didn't know how well this idea would fit into the `zip` crate and you've helped so much to shepherd it into something really useful ^_^!

Writing down TODOs for myself:
- [ ] Implement `info` and `extract`, and gate them by feature flags (https://github.com/zip-rs/zip2/pull/235#discussion_r1717767590).
- [ ] Implement a `WrapErr` trait and `Error` impl that doesn't call into `eyre` or `anyhow`, since we're not making a large server application which may error out in unexpected ways (https://github.com/zip-rs/zip2/pull/235#discussion_r1720810911).
    - Also make CLI arg parsing errors use this error framework instead of calling `process::exit()` by hand.
- [ ] Make errors of the same type produce the same error message.
    - e.g. for `-f` args and positional file args that don't exist, don't write out different error messages, but make them produce the same case of e.g. a `CompressError` struct, which is converted into a standard stderr output + exit code at the top-level `main()` method.


Going to also explicitly set this into draft mode as well and will ping you when it's ready for review again.

cosmicexplorer commented:
Was able to remove the `eyre` dependency by creating our own `WrapErr` trait, and knocked off 300kb from `zip-clite`! The current sizes are 1.4M for `zip-cli` and 481k for `zip-clite`:
```shell
> ls target/release/zip-cli                                                              
-rwxr-xr-x 2 cosmicexplorer wheel 1.4M Aug 17 17:04 target/release/zip-cli*
> ls target/release/zip-clite
-rwxr-xr-x 2 cosmicexplorer wheel 481k Aug 17 17:04 target/release/zip-clite*
```

cosmicexplorer commented:
Not really ready for review yet, as I've still got `info` and `extract` to work on, but I've made a few changes:
1. Moved the `main()` method logic for `zip-cli` and `zip-clite` into `zip_cli::driver::main()` to avoid duplicating that code in both `main.rs` files.
2. Added a `--append` flag for `compress` which opens the `-o` path with `ZipWriter::new_append()` if it already exists, or just creates it new if not.
3. Renamed `CompressError` to `CommandError` so we can reuse it and the error wrapping logic for the other two subcommands.

cosmicexplorer commented:
This got way out of scope, which was still worth it to me, but makes it difficult to justify posting all of this as the first draft of this functionality. I'm going to rebase, cut out the entry matching system (which I think is a great idea but not appropriate for v0), then mark this ready for review.

cosmicexplorer commented:
Hmmmm. I took another look over this and I actually think what's here is very useful, but that it's difficult to understand how to correctly use any of the commands because the `--help` output provides exactly one view of the help text instead of having a brief vs detailed output. I also think structured (json) input would be very very useful for programmatic invocation, instead of having to translate to a command line. I have an idea of how to achieve this, and am working on that now.

The result should be that:
- simple stuff is simple
- command input is decoupled from argument parsing
- `--help` output has a simple and a detailed view

Pr0methean commented:
@cosmicexplorer Could you please either give an update on your plans to finish this PR, or close it? I'm trying to reduce the number of open PRs I have to keep track of.

cosmicexplorer commented:
This PR can totally be redone and scoped way down to the original proposal. I'm going to focus on #236 first now that I have a solution to allocating output file handles, but this is coming next. I'm hoping to make progress on #236 this week, and vaguely expect to be able to prototype this one next week.

cosmicexplorer commented:
Beginning work on this again. I'm going to refocus on the simplest possible interface (which will at start be the same between `zip-cli` and `zip-clite`), vaguely patterned after the `zip` and `unzip` tools:

- `compress`: accept input strictly over lines of stdin.
    - if it ends in `/`, it's a directory (no i/o to check this, just write a directory entry with that name).
    - if not, it's assumed to be a file (symlinks always expanded), and the file contents are read from the given path into the zip.
        - if the file doesn't exist, an error is signaled and execution stops.
    - no configuration of compression behavior, just use the defaults.
- `extract all` (subcommand of `extract`): accepts a single positional zip file path argument.
    - if the zip file path does not exist or produces any error during extraction, print the error to stderr and exit with failure.
    - extract the contents of the zip archive into the current directory.
- `extract single`: accepts a positional zip file path argument, and a positional entry name argument.
    - if the zip file path does not exist or produces any error during loading, print the error to stderr and exit with failure.
    - if the entry name does not exist or produces any error during extraction, print the error to stderr and exit with failure.
    - write the decompressed contents of the entry to stdout.
- `info`: accepts a single positional zip file path argument.
    - if the zip file path does not exist or produces any error during loading, print the error to stderr and exit with failure.
    - prints the entry name list to stdout, in order.

This is possibly the simplest CLI I could think of in order to avoid any further equivocation about parsing command-line arguments. This interface is obviously going to be unstable, but it will let us verify that we can produce binary executables which can interface with the zip file format.

The distinction between `zip-cli` and `zip-clite` will remain meaningful, because `zip-clite` will only support DEFLATE. This should be easy to test by generating test input with the full library API.

I'm going to reuse this PR, because you provided some very thoughtful comments on it, which I appreciated immensely and don't want to lose by closing this. [b0cfa9dba](https://github.com/cosmicexplorer/zip2/commit/b0cfa9dba1391e012f958b8e589a03d88c204432) remains on my fork. Follow-up work can look to these as reference material.
